Improve TS code by adding generics and removing `any` where applicable

ltetzlaff opened the pull request:
This is in response to: https://github.com/sindresorhus/ow/pull/68#issuecomment-385262882

The switch at the top is debatable but should be faster and is also shorter 🙆
Generally it is more an improvement for TS-users and we directly see if our typings are not set correctly, as to be seen in the tests for instance 🕵️
(If you're also interested why `() => {}` is ArrayLike (=== has a length) check this out: [Function.length](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/length))

Please bear with me for throwing plenty of those `any`s in in the first place: [Blame](https://github.com/sindresorhus/is/blame/master/source/index.ts) 🙈🙉

sindresorhus commented:
What was the original argument for using `any` in all of the methods? I remember we had a discussion about it and ended up with that. @SamVerschueren Do you remember/

sindresorhus commented on source/index.ts:
```diff
@@ -118,40 +113,40 @@ namespace is { // tslint:disable-line:no-namespace
 	export const function_ = isOfType<Function>('function');
 	export const null_ = (value: any): value is null => value === null;
```
`any` => `null`?

ltetzlaff commented on source/index.ts:
```diff
@@ -246,22 +241,22 @@ namespace is { // tslint:disable-line:no-namespace
 
 	export const nodeStream = (value: any) => !nullOrUndefined(value) && isObject(value) && function_(value.pipe);
```
`plainObject` and `nodeStream` definitely!
`domElement` is a little tricky since it would be `object & { nodeType: 1, nodeName: string, etc.. }`

sindresorhus replied:
> domElement is a little tricky since it would be object & { nodeType: 1, nodeName: string, etc.. }

But even just `object` (without the `nodeType` etc) for `domElement` would be better than `any`, right?

ltetzlaff commented:
Basically you can still do that if you're a JS user, for TS people that won't really help because the compiler will tell them if what they're trying to do is not allowed and will most likely be used for people to check `CLI inputs`, `JSON.parse()` etc. 

For those cases I'd say working with `const a = JSON.parse("whatever")` which is typed as `any` still allows throwing them into our methods to further decide what we have there.

sindresorhus commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
 	export const function_ = isOfType<Function>('function');
-	export const null_ = (value: any): value is null => value === null;
+	export const null_ = (value: null | any): value is null => value === null;
 	export const class_ = (value: any) => function_(value) && value.toString().startsWith('class ');
```
Could this one have a `class` type?

See:
- https://stackoverflow.com/questions/39392853/is-there-a-type-for-class-in-typescript-and-does-any-include-it
- https://stackoverflow.com/questions/39976329/typescript-type-to-represent-any-class

sindresorhus commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
+	export const nullOrUndefined = (value: null | undefined | any): value is null | undefined => null_(value) || undefined(value);
+	export const object = (value: object) => !nullOrUndefined(value) && (function_(value) || isObject(value));
+	export const iterable = <T = any>(value: IterableIterator<T>): value is IterableIterator<T> => !nullOrUndefined(value) && function_(value[Symbol.iterator]);
+	export const generator = (value: Generator): value is Generator => iterable(value as IterableIterator<any>) && function_(value.next) && function_(value.throw);
```
What's the point of using `value is Generator` when we already define `value: Generator` as a parameter?

SamVerschueren replied:
That might be one of the reasons I went for `(value: any)`. If you already know the type, what's the point in using `is`. It could lead to weird implementations

```ts
const x = some3rdPartyCall();

// Cast it because otherwise we can't test it
if (is.generator(x as Generator)) {
   // Now we know at runtime that `x` is a generator
}
```

sindresorhus commented on source/index.ts:
```diff
@@ -172,7 +167,7 @@ namespace is { // tslint:disable-line:no-namespace
 	export const truthy = (value: any) => Boolean(value);
```
Same as https://github.com/sindresorhus/is/pull/49/files?w=1#r185387543 for `directInstanceOf`'s `klass` argument. (I can't comment on the exact line.

sindresorhus commented on source/index.ts:
```diff
@@ -241,27 +236,27 @@ namespace is { // tslint:disable-line:no-namespace
-	const isEmptyStringOrArray = (value: any) => (string(value) || array(value)) && value.length === 0;
+	const isWhiteSpaceString = (value: string) => string(value) && /\S/.test(value) === false;
+	const isEmptyStringOrArray = (value: string | any[]) => (string(value) || array(value)) && value.length === 0;
 	const isEmptyObject = (value: any) => !map(value) && !set(value) && object(value) && Object.keys(value).length === 0;
```
`any` => `object`?

sindresorhus commented on source/index.ts:
```diff
@@ -241,27 +236,27 @@ namespace is { // tslint:disable-line:no-namespace
+	const isWhiteSpaceString = (value: string) => string(value) && /\S/.test(value) === false;
+	const isEmptyStringOrArray = (value: string | any[]) => (string(value) || array(value)) && value.length === 0;
 	const isEmptyObject = (value: any) => !map(value) && !set(value) && object(value) && Object.keys(value).length === 0;
 	const isEmptyMapOrSet = (value: any) => (map(value) || set(value)) && value.size === 0;
```
`any` => `Map<any, any> | Set<any>`?

SamVerschueren commented:
> What was the original argument for using any in all of the methods? 

Not sure anymore, don't think there was a specific reason.

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
 	export const number = isOfType<number>('number');
 	export const function_ = isOfType<Function>('function');
-	export const null_ = (value: any): value is null => value === null;
+	export const null_ = (value: null | any): value is null => value === null;
```
Why `any`?

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
 
 namespace is { // tslint:disable-line:no-namespace
-	const isObject = (value: any) => typeof value === 'object';
+	const isObject = (value: object) => typeof value === 'object'; // tslint:disable-line:strict-type-predicates
```
Why disabling `strict-type-predicates`? Tested it locally and doesn't seem to fail without disabling it.

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
-	export const object = (value: any) => !nullOrUndefined(value) && (function_(value) || isObject(value));
-	export const iterable = (value: any): value is Iterator<any> => !nullOrUndefined(value) && function_(value[Symbol.iterator]);
-	export const generator = (value: any): value is Generator => iterable(value) && function_(value.next) && function_(value.throw);
+	export const nullOrUndefined = (value: null | undefined | any): value is null | undefined => null_(value) || undefined(value);
```
Same as above, why `any`?

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
-	export const generator = (value: any): value is Generator => iterable(value) && function_(value.next) && function_(value.throw);
+	export const nullOrUndefined = (value: null | undefined | any): value is null | undefined => null_(value) || undefined(value);
+	export const object = (value: object) => !nullOrUndefined(value) && (function_(value) || isObject(value));
+	export const iterable = <T = any>(value: IterableIterator<T>): value is IterableIterator<T> => !nullOrUndefined(value) && function_(value[Symbol.iterator]);
```
We discussed this and we decided not to do this. I explain [here](https://github.com/sindresorhus/is/pull/43#issuecomment-377780830) why.

> I don't think we should do that. The only thing that is does, is check the type and infer the type correctly. It's not up to is to do more than that. If users type there values correctly, all of it will go automatically.

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
+	export const generator = (value: Generator): value is Generator => iterable(value as IterableIterator<any>) && function_(value.next) && function_(value.throw);
 
-	export const nativePromise = isObjectOfType<Promise<any>>(TypeName.Promise);
+	export const nativePromise = <T = any>(value: Promise<T>): value is Promise<T> =>
```
Same as above, remove the generic.

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
+		isObjectOfType<Promise<T>>(TypeName.Promise)(value);
 
-	const hasPromiseAPI = (value: any): value is Promise<any> =>
+	const hasPromiseAPI = <T = any>(value: Promise<T>): value is Promise<T> =>
```
Same as above, remove the generic.

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
 		function_(value.catch);
 
-	export const promise = (value: any): value is Promise<any> => nativePromise(value) || hasPromiseAPI(value);
+	export const promise = <T = any>(value: Promise<T>): value is Promise<T> => nativePromise<T>(value) || hasPromiseAPI<T>(value);
```
Same as above, remove the generic.

SamVerschueren commented on source/index.ts:
```diff
@@ -109,49 +104,49 @@ function is(value: any): TypeName { // tslint:disable-line:only-arrow-functions
+	export const map = <T1 = any, T2 = any>(value: Map<T1, T2>): value is Map<T1, T2> => isObjectOfType<Map<T1, T2>>(TypeName.Map)(value);
+	export const set = <T = any>(value: Set<T>): value is Set<T> => isObjectOfType<Set<T>>(TypeName.Set)(value);
+	export const weakMap = <T1 extends object = any, T2 = any>(value: WeakMap<T1, T2>): value is WeakMap<T1, T2> => isObjectOfType<WeakMap<T1, T2>>(TypeName.WeakMap)(value);
+	export const weakSet = <T extends object = any>(value: WeakSet<T>): value is WeakSet<T> => isObjectOfType<WeakSet<T>>(TypeName.WeakSet)(value);
```
All of these `map`, `set`, `weakMap`, `weakSet` should not use generics. See above for details.

SamVerschueren commented on source/index.ts:
```diff
@@ -241,27 +236,27 @@ namespace is { // tslint:disable-line:no-namespace
 
-	type ArrayMethod = (fn: (value: any, index: number, array: any[]) => boolean, thisArg?: any) => boolean;
-	const predicateOnArray = (method: ArrayMethod, predicate: any, values: any[]) => {
+	type ArrayMethod = <T>(fn: (value: T, index: number, array: T[]) => boolean, thisArg?: any) => boolean;
```
Same as above, remove the generic.

SamVerschueren commented on source/tests/test.ts:
```diff
@@ -582,15 +582,15 @@ test('is.plainObject', t => {
+	t.false(m.iterable(0 as any));
+	t.false(m.iterable(NaN as any));
+	t.false(m.iterable(Infinity as any));
+	t.false(m.iterable({} as any));
```
As explained above, this is the reason why I'm not sure if we should strongly type the `value` arguments. My gut feeling says `I just want to throw values to the library, and it should either return true or false`. Always having to typecast it to the specific type either assumes you already know what it is, or otherwise you need type it as `any`.

Another example would be a function which either returns a `Promise` or an `Observable`.

```ts
const value = someFunction();

// `value` is of type `Promise` | `Observable`, but have to cast it explicitely to `Promise` or any.
if (is.promise(value as Promise)) {
   // ...
}
```

ltetzlaff commented:
According to @SamVerschueren's review I refactored to use `(value: any)` consistently, only `inRange` and `isAbsoluteMod2` are still typed to expect numbers because we do not check every parameter for `number(v)` explicitly, ultimately based on these two very valid arguments:

https://github.com/sindresorhus/is/pull/49#discussion_r185393242
https://github.com/sindresorhus/is/pull/49#discussion_r185392646

To improve the type safety of the result I added `(): value is T` where missing, though.

sindresorhus commented:
Thanks again for such great work @ltetzlaff :)
